packages/aragon-wrapper/src/utils/AsyncRequestCache.js

```diff
@@ -0,0 +1,46 @@
+/**
+ * A cache to deduplicate async requests.
+ */
+export default class AsyncRequestCache {
+  #cache = new Map()
+  #requestFn
+
+  /**
+   * Create a new AsyncRequestCache that will use `requestFn` when requesting each key
+   *
+   * @param  {function} requestFn Async function for requesting each key
+   */
+  constructor (requestFn) {
+    this.#requestFn = requestFn
+  }
+
+  /**
+   * Check if the `key` is available in the cache
+   *
+   * @param  {string} key Key to check
+   * @return {boolean} If key is in the cache
+   */
+  has (key) {
+    return this.#cache.has(key)
+  }
+
+  /**
+   * Request `key`, using previous result if cached.
+   * Resets the cache for `key` if the request was not successful.
+   *
+   * @param  {string} key Key to request
+   * @return {Promise<*>} Request result
+   */
+  request (key) {
+    if (this.has(key)) {
+      return this.#cache.get(key)
+    }
+    const request = Promise.resolve(this.#requestFn(key))
+      .catch((err) => {
+        this.#cache.delete(key)
+        throw err
+      })
+    this.#cache.set(key, request)
```

Comment:
This will cache the promise rather than the resolved value. Is that intentional?

Reply:
>  * A cache to deduplicate async requests.

My bad, this is to deduplicate requests of a key.

Reply:
Yes, the idea is to have a familiar interface, so that it _feels_ exactly the same as if you'd be calling it without the cache.

Reply:
Yeah makes total sense. It's also interoperable with just values (as in the case of `applicationInfoCache` calling `getAragonOsInternalAppInfo`) which is nice